corehq/apps/locations/fixtures.py

```diff
@@ -70,76 +77,98 @@ def __call__(self, user, version, last_sync=None, app=None):
         """
         if not user.project.uses_locations:
             return []
-        if toggles.SYNC_ALL_LOCATIONS.enabled(user.domain):
-            locations = SQLLocation.active_objects.filter(domain=user.domain)
-        else:
-            locations = []
-            user_location = user.sql_location
-            if user_location:
-                # add users location (and ancestors) to fixture
-                locations.append(user_location)
-
-                # add all descendants as well
-                locations += (user_location.get_descendants()
-                                           .filter(is_archived=False))
-
-            if user.project.supports_multiple_locations_per_user:
-                # this might add duplicate locations but we filter that out later
-                location_ids = [loc.location_id for loc in user.locations]
-                locations += SQLLocation.active_objects.filter(
-                    location_id__in=location_ids
-                )
-
-        location_db = _location_footprint(locations)
-
-        if not should_sync_locations(last_sync, location_db, user):
-            return []
 
-        root = Element('fixture',
-                       {'id': self.id,
-                        'user_id': user.user_id})
+        all_locations = _all_locations(user)
 
-        root_locations = filter(
-            lambda loc: loc.parent is None, location_db.by_id.values()
-        )
+        if not should_sync_locations(last_sync, all_locations, user):
+            return []
+
+        root_node = Element('fixture', {'id': self.id, 'user_id': user.user_id})
+        root_locations = all_locations.root_locations
 
         if root_locations:
-            _append_children(root, location_db, root_locations)
-        return [root]
+            _append_children(root_node, all_locations, root_locations)
+        return [root_node]
 
 
 location_fixture_generator = LocationFixtureProvider()
 
 
-def _valid_parent_type(location):
-    parent = location.parent
-    parent_type = parent.location_type if parent else None
-    return parent_type == location.location_type.parent_type
+def _all_locations(user):
+    if toggles.SYNC_ALL_LOCATIONS.enabled(user.domain):
+        return LocationSet(SQLLocation.active_objects.filter(domain=user.domain))
+    else:
+        all_locations = set()
+
+        user_location = user.sql_location
+        user_locations = set([user_location]) if user_location is not None else set()
+        user_locations |= {location for location in _gather_multiple_locations(user)}
```

Comment:
nit: you can also do `set(_gather_multiple_locations(user))` here to coerce the generator to a set